### PR TITLE
EbAppConfig: Support double dashes for hme 

### LINKED ISSUE #441: If scene change detection is enabled, it can not encode normally.
SVT-AV1 8977f44

input video
https://drive.google.com/open?id=17lh4I8wZ1rELRY04ysOz7kxDWF-hIfvh

output video
https://drive.google.com/open?id=12eWol2s-yyxIQxeDYlE8BgzyrTFw5WBw

Commandline
``SvtAv1EncApp.exe -i Big_Buck_Bunny_scd1.y4m -enc-mode 8 -intra-period 23 -scd 1 -q 30 -n 240 -b Big_Buck_Bunny_scd1.ivf``

（c）copyright 2008、Blender Foundation / www.bigbuckbunny.org

### LINKED ISSUE #1271: Confused about 'Unprocessed tokens'
calling:
```
ffmpeg -y -loglevel fatal -noautorotate -nostdin -threads 8 -i "F:\TestClips&Co\files\test.avi" -map 0:0 -an -sn -vf zscale=rangein=tv:range=tv -pix_fmt yuv420p10le -strict -1 -vsync 0  -f rawvideo - | SvtAv1EncApp --qp 50 --use-default-me-hme 0 --search-w 256 --search-h 256 --hme-tot-l0-w 256 --hme-tot-l0-h 256 --hme-l0-w 256 --hme-l0-h 256 --hme-l1-w 256 --hme-l1-h 256 --hme-l2-w 256 --hme-l2-h 256  --input stdin  -n 429 --width 640 --height 352 --fps-num 25 --fps-denom 1 --input-depth 10 --output "E:\Temp\17_57_36_8410_01.av1"
```
I get:
```
Unprocessed tokens:  --hme-l0-w  --hme-l2-h  --hme-l0-h  --hme-l1-w  --hme-l1-h  --hme-l2-w

Error in configuration, could not begin encoding! ...
Run SvtAv1EncApp -help for a list of options
Encoder finished
```
But the help states:
```
--hme-l0-w                	Set hierarchical motion estimation level0 search area in Width
--hme-l0-h                	Set hierarchical motion estimation level0 search area in height
--hme-l1-w                	Set hierarchical motion estimation level1 search area in Width
--hme-l1-h                	Set hierarchical motion estimation level1 search area in height
--hme-l2-w                	Set hierarchical motion estimation level2 search area in Width
--hme-l2-h                	Set hierarchical motion estimation level2 search area in height
```
So why are those setting not processed?

### PR DESCRIPTION
**Pull request by 1480c1**
Closes #1271
Closes #441 

Removed `--scd` since it was broken for a long time

Changed printed out help text to double dash
Corrected asm help text in the encoder guide along with replacing single dash long options with their correct ones

### CODE REVIEW

File: Source/App/EncApp/EbAppConfig.c [hunk 1]
```diff
@@ -24,52 +24,52 @@
...
+#define ERROR_FILE_TOKEN "--errlog"
 #define QP_FILE_TOKEN "-qp-file"
-#define INPUT_STAT_FILE_TOKEN "-input-stat-file"
-#define OUTPUT_STAT_FILE_TOKEN "-output-stat-file"
-#define STAT_FILE_TOKEN "-stat-file"
-#define INPUT_PREDSTRUCT_FILE_TOKEN "-pred-struct-file"
+#define INPUT_STAT_FILE_TOKEN "--input-stat-file"
+#define OUTPUT_STAT_FILE_TOKEN "--output-stat-file"
+#define STAT_FILE_TOKEN "--stat-file"
+#define INPUT_PREDSTRUCT_FILE_TOKEN "--pred-struct-file"
```
Comment (tianjunwork) [inline Source/App/EncApp/EbAppConfig.c:39]: please also update user guide.

File: Source/App/EncApp/EbAppConfig.c [hunk 2]
```diff
@@ -24,52 +24,52 @@
...
 #define WIDTH_TOKEN "-w"
 #define HEIGHT_TOKEN "-h"
 #define NUMBER_OF_PICTURES_TOKEN "-n"
-#define BUFFERED_INPUT_TOKEN "-nb"
+#define BUFFERED_INPUT_TOKEN "--nb"
 #define NO_PROGRESS_TOKEN "--no-progress"
-#define BASE_LAYER_SWITCH_MODE_TOKEN "-base-layer-switch-mode" // no Eval
+#define BASE_LAYER_SWITCH_MODE_TOKEN "--base-layer-switch-mode" // no Eval
 #define QP_TOKEN "-q"
```
Comment (tianjunwork) [inline Source/App/EncApp/EbAppConfig.c:46]: `-q` is already `--qp` in user guide.

File: Docs/svt-av1_encoder_user_guide.md [hunk 3]
```diff
@@ -124,10 +124,11 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
 | --- | --- | --- | --- | --- |
 | **InputFile** | -i | any string | None | Input file path; "stdin" to take input from standard input stream |
 | **StreamFile** | -b | any string | null | Output bitstream file path; "stdout" to emit output to standard output stream |
-| **ConfigFile** | -c | any string | null | Configuration file path |
+| **ConfigFile** | -c,  | any string | null | Configuration file path |
```
Comment (tianjunwork) [inline Docs/svt-av1_encoder_user_guide.md:127]: extra `,` not needed

File: Docs/svt-av1_encoder_user_guide.md [hunk 4]
```diff
@@ -124,10 +124,11 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
...
 | **ErrorFile** | --errlog | any string | stderr | error log displaying configuration or encode errors |
 | **ReconFile** | -o | any string | null | Recon file path. Optional output of recon. |
 | **StatFile** | --stat-file | any string | Null | Path to statistics file if specified and StatReport is set to 1, per picture statistics are outputted in the file|
+| **PredStructFile** | [none] | any string | Null | Manual prediction structure file path, refer to the comments in the `Config/PredStruct_level{0..5}.cfg` for specific details|
```
Comment (tianjunwork) [inline Docs/svt-av1_encoder_user_guide.md:131]: what is the option to use here?

File: Docs/svt-av1_encoder_user_guide.md [hunk 5]
```diff
@@ -139,6 +140,7 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
 | **BufferedInput** | --nb | [-1, 1 to 2^31 -1] | -1 | number of frames to preload to the RAM before the start of the encode If --nb = 100 and -n 1000 -- > the encoder will encode the first 100 frames of the video 10 times |
 | **EncoderColorFormat** | --color-format | [1 for default] | 1 | Set encoder color format(EB_YUV400, EB_YUV420, EB_YUV422, EB_YUV444) |
 | **Profile** | --profile | [0-2, 0 for default] | 0 | Bitstream profile number to use (0: main profile[default], 1: high profile, 2: professional profile) |
+| **Tier** | [none] | [0-1] | 0 | Constraints for bitstream in terms of max bitrate and max buffer size (0: main[default], 1: high) |
```
Comment (tianjunwork) [inline Docs/svt-av1_encoder_user_guide.md:143]: same here. what is the option to use?

File: Docs/svt-av1_encoder_user_guide.md (hunk 4 above)
Comment (1480c1) [reply Docs/svt-av1_encoder_user_guide.md:131]: This option is exposed in the config files, but not in the command line, it seems it has a token allocated to it, but not an actual slot `--pred-struct-file`

File: Docs/svt-av1_encoder_user_guide.md (hunk 3 above)
Comment (1480c1) [reply Docs/svt-av1_encoder_user_guide.md:127]: typo, will remove

File: Docs/svt-av1_encoder_user_guide.md (hunk 5 above)
Comment (1480c1) [reply Docs/svt-av1_encoder_user_guide.md:143]: Similar to above, the option is exposed in the config files, but isn't allocated a slot in the command line `--tier`

File: Source/App/EncApp/EbAppConfig.c [hunk 6]
```diff
@@ -24,52 +24,52 @@
...
-#define INPUT_PREDSTRUCT_FILE_TOKEN "-pred-struct-file"
+#define INPUT_STAT_FILE_TOKEN "--input-stat-file"
+#define OUTPUT_STAT_FILE_TOKEN "--output-stat-file"
+#define STAT_FILE_TOKEN "--stat-file"
+#define INPUT_PREDSTRUCT_FILE_TOKEN "--pred-struct-file"
 #define WIDTH_TOKEN "-w"
 #define HEIGHT_TOKEN "-h"
 #define NUMBER_OF_PICTURES_TOKEN "-n"
-#define BUFFERED_INPUT_TOKEN "-nb"
+#define BUFFERED_INPUT_TOKEN "--nb"
```
Comment (hassount) [inline Source/App/EncApp/EbAppConfig.c:43]: would we still have backward compatibility with the old tokens if we do this ?

Comment (1480c1) [reply Source/App/EncApp/EbAppConfig.c:43]: Does not seem so

I am trying in another branch to see what would be the best way to use double dash while keeping single dash compatibility

Comment (1480c1) [reply Source/App/EncApp/EbAppConfig.c]: Reverted the changes and made it so it only affects them stuff

**Comment by ccccheung** [comment]
- This PR closes #1271 and now processes the tokens mentioned in the issue
- However, I encountered another issue with HME parameters; for parameters 
``` --hme-l0-w 256 --hme-l0-h 256 --hme-l1-w 256 --hme-l1-h 256 --hme-l2-w 256 --hme-l2-h 256``` 
 the encoder prints Invalid  HME Total Search Area. Must be [1 - 256]
However the value provided is 256

File: Source/App/EncApp/EbAppConfig.c [hunk 7]
```diff
@@ -1243,10 +1244,6 @@ ConfigEntry config_entry[] = {
     {SINGLE_INPUT, TILE_ROW_TOKEN, "TileRow", set_tile_row},
     {SINGLE_INPUT, TILE_COL_TOKEN, "TileCol", set_tile_col},
     // Rate Control
-    {SINGLE_INPUT,
```
Comment (hassount) [inline Source/App/EncApp/EbAppConfig.c:1246]: I would recommend not removing the code for -scd rather only accept the 0 state for now as we would ultimately have to support it.

Comment (1480c1) [reply Source/App/EncApp/EbAppConfig.c:1246]:
```c
static void set_scene_change_detection(const char *value, EbConfig *cfg) {
    cfg->scene_change_detection = 0;
}
```
Then should I do something like this instead? Allow it to trigger, but only set it to 0

Comment (hassount) [reply Source/App/EncApp/EbAppConfig.c:1246]: I'd rather we allow the user to select 0/1 but reject the -scd 1 option. hard coding it to 0 would give users the impression that it is supported which is not the case. BTW, a nice project for you to work on if you'd like :)

Comment (1480c1) [reply Source/App/EncApp/EbAppConfig.c:1246]: > I'd rather we allow the user to select 0/1 but reject the -scd 1 option

We can do that by modifying the regular check for scene detection argument



> BTW, a nice project for you to work on if you'd like :)

I do have a slight interest in scene detection for one reason or another

Comment (1480c1) [reply Source/App/EncApp/EbAppConfig.c:1246]: `"Scene change detection is currently broken and must be set to 0"`
or
`"Scene change detection must be 0"`
Should we include a message as to why it is enforced to 0, or no?

Comment (hassount) [reply Source/App/EncApp/EbAppConfig.c:1246]: how about '''"Scene change detection is currently not supported"''' ?

Comment (1480c1) [reply Source/App/EncApp/EbAppConfig.c]: Short and simple, that's good

**Comment by 1480c1** [comment]
> * This PR closes #1271 and now processes the tokens mentioned in the issue
> * However, I encountered another issue with HME parameters; for parameters
>   ` --hme-l0-w 256 --hme-l0-h 256 --hme-l1-w 256 --hme-l1-h 256 --hme-l2-w 256 --hme-l2-h 256`
>   the encoder prints Invalid  HME Total Search Area. Must be [1 - 256]
>   However the value provided is 256

The reason why it fails is that the `--hme-l0-w` parameter requires two (or whatever is passed to `--num-hme-w`) values that equal 256 when combined rather than just one. So something like `--hme-tot-l0-w 480 --hme-l0-w 240 240` where 240+240==480

**Comment by Miraculin** [comment]
> > * This PR closes #1271 and now processes the tokens mentioned in the issue
> > * However, I encountered another issue with HME parameters; for parameters
> >   ` --hme-l0-w 256 --hme-l0-h 256 --hme-l1-w 256 --hme-l1-h 256 --hme-l2-w 256 --hme-l2-h 256`
> >   the encoder prints Invalid  HME Total Search Area. Must be [1 - 256]
> >   However the value provided is 256
> 
> The reason why it fails is that the `--hme-l0-w` parameter requires two (or whatever is passed to `--num-hme-w`) values that equal 256 when combined rather than just one. So something like `--hme-tot-l0-w 480 --hme-l0-w 240 240` where 240+240==480

I've re-run the encoding using this information
- the new upper limit of 480 is respected
- as long as the sum is exactly the value of ```--hme-tot-l0-w``` and ```--hme-tot-l0-h``` respectively, the encoding runs
- Giving 2 values that don't add up to ```hme-tot-l0-w/hme-tot-l0-h``` as well as only providing 1 value gives the error : ```Invalid HME Total Search Area``` as expected. While the answer to what went wrong is found in the help text, should a hint be provided here in the error text about what proper arguments are?

**Comment by ccccheung** [comment]
> > * This PR closes #1271 and now processes the tokens mentioned in the issue
> > * However, I encountered another issue with HME parameters; for parameters
> >   ` --hme-l0-w 256 --hme-l0-h 256 --hme-l1-w 256 --hme-l1-h 256 --hme-l2-w 256 --hme-l2-h 256`
> >   the encoder prints Invalid  HME Total Search Area. Must be [1 - 256]
> >   However the value provided is 256
> 
> The reason why it fails is that the `--hme-l0-w` parameter requires two (or whatever is passed to `--num-hme-w`) values that equal 256 when combined rather than just one. So something like `--hme-tot-l0-w 480 --hme-l0-w 240 240` where 240+240==480

I re-ran the encodings using this information
- the new upper limit of 480 is respected
- Giving 2 parameters that sum exactly to `--hme-tot-l0-w` and `--hme-tot-l0-h` respectively worked as expected.
- Giving 2 parameters that don't don't sum or providing the wrong number of parameters from `-num-hme-w/-num-hme-h` gives `Invalid HME total search area` as expected. While the help text covers these requirements, should a hint be provided here in the error about what went wrong?

**Comment by 1480c1** [comment]
> > > * This PR closes #1271 and now processes the tokens mentioned in the issue
> > > * However, I encountered another issue with HME parameters; for parameters
> > >   ` --hme-l0-w 256 --hme-l0-h 256 --hme-l1-w 256 --hme-l1-h 256 --hme-l2-w 256 --hme-l2-h 256`
> > >   the encoder prints Invalid  HME Total Search Area. Must be [1 - 256]
> > >   However the value provided is 256
> > 
> > 
> > The reason why it fails is that the `--hme-l0-w` parameter requires two (or whatever is passed to `--num-hme-w`) values that equal 256 when combined rather than just one. So something like `--hme-tot-l0-w 480 --hme-l0-w 240 240` where 240+240==480
> 
> I re-ran the encodings using this information
> 
> * the new upper limit of 480 is respected
> * Giving 2 parameters that sum exactly to `--hme-tot-l0-w` and `--hme-tot-l0-h` respectively worked as expected.
> * Giving 2 parameters that don't don't sum or providing the wrong number of parameters from `-num-hme-w/-num-hme-h` gives `Invalid HME total search area` as expected. While the help text covers these requirements, should a hint be provided here in the error about what went wrong?

What do you think would be a good hint text? `"Summed values of hme area does not equal total area"`?

**Comment by ccccheung** [comment]
> > > > * This PR closes #1271 and now processes the tokens mentioned in the issue
> > > > * However, I encountered another issue with HME parameters; for parameters
> > > >   ` --hme-l0-w 256 --hme-l0-h 256 --hme-l1-w 256 --hme-l1-h 256 --hme-l2-w 256 --hme-l2-h 256`
> > > >   the encoder prints Invalid  HME Total Search Area. Must be [1 - 256]
> > > >   However the value provided is 256
> > > 
> > > 
> > > The reason why it fails is that the `--hme-l0-w` parameter requires two (or whatever is passed to `--num-hme-w`) values that equal 256 when combined rather than just one. So something like `--hme-tot-l0-w 480 --hme-l0-w 240 240` where 240+240==480
> > 
> > 
> > I re-ran the encodings using this information
> > 
> > * the new upper limit of 480 is respected
> > * Giving 2 parameters that sum exactly to `--hme-tot-l0-w` and `--hme-tot-l0-h` respectively worked as expected.
> > * Giving 2 parameters that don't don't sum or providing the wrong number of parameters from `-num-hme-w/-num-hme-h` gives `Invalid HME total search area` as expected. While the help text covers these requirements, should a hint be provided here in the error about what went wrong?
> 
> What do you think would be a good hint text? `"Summed values of hme area does not equal total area"`?

Sounds good to me

**Comment by ccccheung** [comment]
latest commit does provide the hint when 2 numbers don't sum up exactly to `--hme-tot-l0-w` or `--hme-tot-l0-h`  in both under sum and over sum, and encoder runs as expected when they do sum exactly

**Comment by hassount** [comment]
@adaria 
is this good to go ?

**Comment by adaria** [comment]
> @adaria
> is this good to go ?

Approved, thank you for the nudge!

**Pull request merged**